sql/hive/src/main/scala/org/apache/spark/sql/hive/HiveExternalCatalog.scala

```diff
@@ -1175,6 +1205,27 @@ private[spark] class HiveExternalCatalog(conf: SparkConf, hadoopConf: Configurat
     client.listFunctions(db, pattern)
   }
 
+  /** Detect whether a table is stored with Hive-compatible metadata. */
+  private def isHiveCompatible(table: CatalogTable): Boolean = {
```

Comment:
Let me revert the question for you: why is it bad to have the compatibility bit be a table property?

You cannot make a non-compatible table compatible after it's been created. At the very least, there is zero code that would achieve that today. And the key to this fix is that the `alterTableSchema` method needs to know whether the table is compatible or not in a reliable way.